[FIX] l10n_es_aeat_sii: Conversion de divisas a moneda de compañia

hugosantosred opened the pull request:
Este PR es para solventar un problema que nos hemos encontrado con la conversión de importes a moneda de compañía.

Al realizar el calculo de precio unitario por linea ya en moneda de compañía y con el redondeo de la conversión aplicado si hay bastantes unidades de un producto estamos perdiendo decimales.

Por ejemplo para una factura con una línea de 9,95$ por ud y 45 ud, teniendo una tasa de cambio de 1,1418:

- Al realizar la conversión de la manera actual obtenemos un precio unitario en moneda de compañía de 8,71€ que multiplicado por las 45 ud nos sale 391,95€.
- Si nos vamos a la información de factura vemos que en el campo "Subtotal moneda compañía" tenemos 392,14€, lo cual nos está descuadrando 19 centimos respecto a la información que tenemos nosotros en la factura y lo que estamos enviando al SII
- Lo mismo nos pasa con el importe de los impuestos, obteniendo un valor de CuotaRepercutida de 82,13€ mientras que en nuestra factura el valor de "Impuesto moneda compañía" es de 82,35, en este caso 22 centimos de diferencia

Con este cambio del PR, al no realizar la solicitud de precio unitario ya redondeado obtenemos:

- Un precio unitario de 8,714310737432124 que multiplicado por las 45 ud. nos da un valor de 392,1439831844456€ que después de redondear son 392,14€ (El mismo valor que en la factura)
- En el valor enviado a SII para la CuotaRepercutida es de 82,35000000000001€, No se por qué no redondea el valor enviado en este caso (Aunque se le aplique el float_round lo mantiene tal cual). 

¿Estaría bien modificarlo y sacar el valor final para enviar utilizando la funcion float_repr del float_utils del ERP? Si me decis que sí lo cambio y lo incluyo en este PR

hugosantosred committed:
[FIX] l10n_es_aeat_sii: Conversion de divisas a moneda de compañia, solventados problemas de decimales

pedrobaeza commented:
El problema que comentas sobre utilizar el `float_round` y que no redondee realmente me lo he encontrado yo también en la v9 y al final opté por utilizar `round(<amount>, 2)`. Por lo demás, me parece correcto el PR (aunque tú dirás si tienes que cambiar lo del redondeo), pero cambia la versión del módulo.

hugosantosred commented:
@pedrobaeza Perfecto, cambio la versión del módulo y hago unas pruebas con el round a ver si sale correcto y actualizo el PR

pedrobaeza commented:
Alguna novedad con esto?

hugosantosred committed:
[FIX] l10n_es_aeat_sii: Utilizar round en conjuncion con float_round

Con esto evitamos los casos en los que float_round no es efectivo por ejemplo hemos encontrado algun caso con valor 82,35000000000001 que utilizando round delante ya convierte correctamente a 82,35

hugosantosred commented:
He aplicado el round delante del float_round para asegurar que van siempre con 2 decimales los valores (No sé si esto será un poco recursivo aunque hay veces que el round de python normal no es todo lo preciso que debería)

También he modificado la versión del módulo

liebana commented:
LGTM :+1:

pedrobaeza reviewed (approved):
Bueno, sí es un poco excesivo tanto round y float_round, pero bueno, como esta versión es marginal y en 
>v9 ya no se requiere tantos round, lo dejo así.
